Sources/GRPCHTTP2Core/GRPCMessageDeframer.swift

```
@@ -0,0 +1,90 @@
+  typealias InboundOut = [UInt8]
+
+  private let decompressor: Zlib.Decompressor?
+  private let maximumPayloadSize: Int?
+  private var effectiveMaximumPayloadSize: Int {
+    self.maximumPayloadSize ?? Self.defaultMaximumPayloadSize
+  }
+
+  /// Create a new ``GRPCMessageDeframer``.
+  /// - Parameters:
+  ///   - maximumPayloadSize: The maximum size a message payload can be.
+  ///   - decompressor: A `Zlib.Decompressor` to use when decompressing compressed gRPC messages.
+  /// - Important: You must call `end()` on the `decompressor` when you're done using it, to clean
+  /// up any resources allocated by `Zlib`.
+  init(maximumPayloadSize: Int? = nil, decompressor: Zlib.Decompressor? = nil) {
+    self.maximumPayloadSize = maximumPayloadSize
+    self.decompressor = decompressor
+  }
+
+  mutating func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
+    guard buffer.readableBytes >= Self.metadataLength else {
+      // If we cannot read enough bytes to cover the metadata's length, then we
+      // need to wait for more bytes to become available to us.
+      return nil
+    }
+
+    // Store the current reader index in case we don't yet have enough
+    // bytes in the buffer to decode a full frame, and need to reset it.
+    let originalReaderIndex = buffer.readerIndex
+    let isMessageCompressed = buffer.readInteger(as: UInt8.self)! == 1
+
+    guard var message = buffer.readLengthPrefixedSlice(as: UInt32.self) else {
+      // We have moved the reader index, but we don't have enough bytes to
+      // read the full message payload, so reset it to its previous, original
+      // position for now, and return. We'll try decoding again, once more
+      // bytes become available in our buffer.
+      buffer.moveReaderIndex(to: originalReaderIndex)
+      return nil
+    }
+
+    if isMessageCompressed {
+      guard let decompressor = self.decompressor else {
+        // We cannot decompress the payload - discard the message.
+        return nil
```

We need to throw an error here: the other side has sent a compressed message but we don't have a compressor. Since compression is negotiated it means that either we, or the remote peer, has a bug.